File: java/src/main/java/dev/intermediatebox/hashtable/InsertDeleteGetRandomO1.java
package dev.intermediatebox.hashtable;

import java.util.*;

public class InsertDeleteGetRandomO1 {
  public static void main(String[] args) {
    RandomizedSet r = new RandomizedSet();

    System.out.println(r.insert(1)); // return true
    System.out.println(r.remove(2)); // return false
    System.out.println(r.insert(2)); // return true
    System.out.println(r.getRandom());  // return random value 1 or 2
    System.out.println(r.remove(1)); // return true
    System.out.println(r.insert(2)); // return false
    System.out.println(r.getRandom()); // always return 2
  }

  // Tag: HashTable

  // O(1) time complexity for all operations
  // O(n) space complexity for hash and map
  private static class RandomizedSet {
    private Map<Integer, Integer> map;
    private List<Integer> list;

    public RandomizedSet() {
      map = new HashMap<>();
      list = new ArrayList<>();
    }

    public boolean insert(int val) {
      if (map.containsKey(val)) {
        return false;
      }

      map.put(val, list.size());
      list.add(val);

      return true;
    }

    public boolean remove(int val) {
      if (!map.containsKey(val)) {
        return false;
      }

      int index = map.get(val);

      if (index < list.size() - 1) {
        // switch last value with value at current index
        int lastValue = list.get(list.size() - 1);
        map.put(lastValue, index);
        list.set(index, lastValue);
      }

      // removes value from the map
      map.remove(val);
      // removes value from the list
      list.remove(list.size() - 1);

      return true;
    }

    public int getRandom() {
      return list.get(new Random().nextInt(list.size()));
    }
  }
}

File: typescript/src/hashtable/InsertDeleteGetRandomO1.ts
(() => {
  class RandomizedSet {
    private map: Map<number, number>;
    private list: Array<number>;

    constructor() {
      this.map = new Map();
      this.list = new Array();
    }

    insert(val: number): boolean {
      if (this.map.has(val)) {
        return false;
      }

      this.map.set(val, this.list.length);
      this.list.push(val);

      return true;
    }

    remove(val: number): boolean {
      if (!this.map.has(val)) {
        return false;
      }

      const index = this.map.get(val)!;

      if (index < this.list.length - 1) {
        const lastValue = this.list[this.list.length - 1];
        this.map.set(lastValue, index);
        this.list[index] = lastValue;
      }

      this.map.delete(val);
      this.list.splice(this.list.length - 2, 1);

      return true;
    }

    getRandom(): number {
      return this.list[Math.floor(Math.random() * this.list.length)];
    }
  }

  const r = new RandomizedSet();
  console.log(r.insert(1)); // return true
  console.log(r.remove(2)); // return false
  console.log(r.insert(2)); // return true
  console.log(r.getRandom()); // return random value 1 o
  console.log(r.remove(1)); // return true
  console.log(r.insert(2)); // return false
  console.log(r.getRandom()); // always return 2
})();